pkg/bpf/events.go

```diff
@@ -0,0 +1,287 @@
+	b := &eventsBuffer{
+		buffer:   container.NewRingBuffer(maxSize),
+		eventTTL: eventsTTL,
+	}
+	if b.eventTTL > 0 {
+		m.scopedLogger().Debug("starting bpf map event buffer GC controller")
+		mapControllers.UpdateController(
+			fmt.Sprintf("bpf-event-buffer-gc-%s", m.name),
+			controller.ControllerParams{
+				DoFunc: func(_ context.Context) error {
+					m.scopedLogger().Debugf("clearing bpf map events older than %s", b.eventTTL)
+					b.buffer.Compact(func(e interface{}) bool {
+						event, ok := e.(*Event)
+						if !ok {
+							log.WithError(wrongObjTypeErr(e)).Error("Failed to compact the event buffer")
+							return false
+						}
+						return time.Since(event.Timestamp) < b.eventTTL
+					})
+					return nil
+				},
+				RunInterval: b.eventTTL,
+			},
+		)
+	}
+	m.events = b
+}
+
+// eventsBuffer stores a buffer of events for auditing and debugging
+// purposes.
+type eventsBuffer struct {
+	buffer        *container.RingBuffer
+	eventTTL      time.Duration
+	subsLock      lock.RWMutex
+	subscriptions []*Handle
+}
+
+// Handle allows for handling event streams safely outside of this package.
+// The key design consideration for event streaming is that it is non-blocking.
+// The eventsBuffer takes care of closing handles when their consumer is not reading
+// off the buffer (or is not reading off it fast enough).
+type Handle struct {
+	c      chan *Event
+	closed *atomic.Value
+	closer *sync.Once
+	err    error
+}
+
+// Returns read only channel for Handle subscription events. Channel should be closed with
+// handle.Close() function.
+func (h *Handle) C() <-chan *Event {
+	return h.c // return read only channel to prevent closing outside of Close(...).
+}
+
+// Close allows for safaley closing of a handle.
+func (h *Handle) Close() {
+	h.close(nil)
+}
+
+func (h *Handle) close(err error) {
+	h.closer.Do(func() {
+		close(h.c)
+		h.err = err
+		h.closed.Store(true)
+	})
+}
+
+func (h *Handle) isClosed() bool {
+	v := h.closed.Load()
+	return v.(bool)
+}
+
+func (h *Handle) isFull() bool {
+	return len(h.c) >= cap(h.c)
+}
+
+// This configures how big buffers are for channels used for streaming events from
+// eventsBuffer.
+//
+// To prevent blocking bpf.Map operations, subscribed events are buffered per client handle.
+// How fast subscribers will need to proceess events will depend on the event throughput.
+// In this case, our throughput will be expected to be not above 100 events a second.
+// Therefore the consumer will have 10ms to process each event. The channel is also
+// given a constant buffer size in the case where events arrive at once (i.e. all 100 events
+// arriving at the top of the second).
+//
+// NOTE: Although using timers/timed-contexts seems like an obvious choice for this use case,
+// the timer.After implementation actually uses a large amount of memory. To reduce memory spikes
+// in high throughput cases, we instead just use a sufficiently buffered channel.
```

Comment:
We have `pkg/inctimer` which I believe resolves the issue with `time.After()`, FWIW.

Reply:
Thats good to know. I think this approach might still be preferable, timers would also introduce the slight question about how long we want to keep the bpf package waiting as we block on a channel. This approach means that bpf will never block on a client subscription.